widgetssdk/src/main/java/com/glia/widgets/chat/ChatManager.kt

```diff
@@ -195,23 +204,84 @@ internal class ChatManager(
             is Action.OperatorConnected -> mapOperatorConnected(action, state)
             Action.Transferring -> mapTransferring(state)
             is Action.OperatorJoined -> mapOperatorJoined(action, state)
-            is Action.UnsentMessageReceived -> addUnsentMessage(action.message, state)
             is Action.ResponseCardClicked -> mapResponseCardClicked(action.responseCard, state)
             is Action.OnMediaUpgradeStarted -> mapMediaUpgrade(action.isVideo, state)
             Action.OnMediaUpgradeToVideo -> mapUpgradeMediaToVideo(state)
             Action.OnMediaUpgradeCanceled -> mapMediaUpgradeCanceled(state)
             is Action.OnMediaUpgradeTimerUpdated -> mapMediaUpgradeTimerUpdated(action.formattedValue, state)
             is Action.CustomCardClicked -> mapCustomCardClicked(action, state)
             Action.ChatRestored, Action.None -> state
-            is Action.MessageClicked -> mapMessageClicked(action.messageId, state)
-            is Action.MessageSent -> mapMessageSent(action.message, state)
-            is Action.MessageSendError -> addErrorMessage(action.message, state)
-            is Action.RemoveMessage -> removeMessage(action.messageId, state)
+            is Action.AttachmentPreviewAdded -> mapAttachmentPreviewAdded(action.attachments, action.payload, state)
+            is Action.MessagePreviewAdded -> mapMessagePreviewAdded(action.visitorChatItem, action.payload, state)
+            is Action.OnMessageSent -> mapMessageSent(action.message, state)
+            is Action.OnSendMessageError -> mapSendMessageFailed(action.messageId, state)
+            is Action.OnRetryClicked -> mapRetryClicked(action.messageId, state)
         }
     }
 
-    @VisibleForTesting
-    fun mapMessageSent(message: VisitorMessage, state: State): State = mapNewMessage(ChatMessageInternal(message), state)
+    private fun mapRetryClicked(messageId: String, state: State): State = state.apply {
+        val payload = messagePreviews[messageId] ?: return@apply
+
+        val files = (payload.attachment as? FilesAttachment)?.files.orEmpty()
+
+        val messageIndex = chatItems.indexOfLast { it.id == payload.messageId }
+
+        if (messageIndex != -1) {
+            chatItems[messageIndex] = (chatItems[messageIndex] as VisitorChatItem).withStatus(VisitorItemStatus.PREVIEW)
+        }
+
+        if (files.isEmpty()) return@apply
+
+        val firstFileIndex = if (messageIndex != -1)
+            messageIndex + 1
+        else
+            chatItems.indexOfFirst { it.id == files.first().id }
+
+        for (index in firstFileIndex until firstFileIndex + files.size) {
+            chatItems[index] = (chatItems[index] as VisitorChatItem).withStatus(VisitorItemStatus.PREVIEW)
+        }
+    }
+
+    private fun mapMessageSent(message: VisitorMessage, state: State): State = mapNewMessage(ChatMessageInternal(message), state)
```

Comment:
Does it make sense to use `mapNewMessage()` directly, instead of a `mapMessageSent()` wrapper?

Reply:
done ✅